core/state.py

```diff
@@ -1136,6 +1136,53 @@ def _MakeArgvCell(argv):
     return Cell(False, False, False, value.List(items))
 
 
+class ctx_Eval(object):
+    """Push temporary variable frame and override $0, $1, $2, etc."""
+
+    def __init__(self, mem, dollar0, pos_args, vars):
+        # type: (Mem, Optional[str], Optional[List[str]], Optional[Dict[str, value_t]]) -> None
+        self.mem = mem
+        self.pushed_pos_args = False
+        self.pushed_dollar0 = False
+        self.pushed_vars = False
+
+        # $0 needs to have lexical scoping. So we store it with other locals.
+        # As "0" cannot be parsed as an lvalue, we can safely store arg0 there.
+        if dollar0 is not None:
+            self.pushed_dollar0 = True
+            assert mem.GetValue("0", scope_e.LocalOnly).tag() == value_e.Undef
+            self.lval = LeftName("0", loc.Missing)
```

Comment:
I might call this `self.dollar0_lval` or something ... `self.lval` is a bit generic